File: packages/typescript/sdk/core/__tests__/store.test.ts
import { describe, expect, test } from "bun:test";
import { exposeStore, type NodeDescriptor, type StateStore, type StoreTarget } from "../src/index";

class TestStore<S> implements StateStore<S> {
  private listeners = new Set<() => void>();

  constructor(private state: S) {}

  getState(): S {
    return this.state;
  }

  subscribe(listener: () => void): () => void {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }

  setState(next: S): void {
    this.state = next;
    for (const listener of [...this.listeners]) {
      listener();
    }
  }

  listenerCount(): number {
    return this.listeners.size;
  }
}

function createTarget() {
  const registered = new Map<string, NodeDescriptor>();
  const unregistered: Array<{ path: string; opts?: { recursive?: boolean } }> = [];

  const target: StoreTarget & {
    registered: Map<string, NodeDescriptor>;
    unregistered: Array<{ path: string; opts?: { recursive?: boolean } }>;
  } = {
    registered,
    unregistered,
    register(path, descriptor) {
      registered.set(path, descriptor);
    },
    unregister(path, opts) {
      registered.delete(path);
      if (opts?.recursive) {
        const prefix = `${path}/`;
        for (const key of [...registered.keys()]) {
          if (key.startsWith(prefix)) registered.delete(key);
        }
      }
      unregistered.push({ path, opts });
    },
  };

  return target;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

describe("exposeStore", () => {
  test("registers the initial store state and updates on store emissions", () => {
    const target = createTarget();
    const store = new TestStore({ count: 1 });

    exposeStore(target, "counter", store, (state) => ({
      type: "status",
      props: { count: state.count },
    }));

    expect(target.registered.get("counter")).toEqual({
      type: "status",
      props: { count: 1 },
    });

    store.setState({ count: 2 });

    expect(target.registered.get("counter")).toEqual({
      type: "status",
      props: { count: 2 },
    });
  });

  test("cleans up the subscription and unregisters recursively", () => {
    const target = createTarget();
    const store = new TestStore({ items: 1 });

    target.registered.set("list/item-1", { type: "item" });

    const cleanup = exposeStore(target, "list", store, (state) => ({
      type: "collection",
      props: { count: state.items },
    }));

    expect(store.listenerCount()).toBe(1);

    cleanup();

    expect(store.listenerCount()).toBe(0);
    expect(target.registered.has("list")).toBe(false);
    expect(target.registered.has("list/item-1")).toBe(false);
    expect(target.unregistered).toContainEqual({ path: "list", opts: { recursive: true } });

    store.setState({ items: 2 });
    expect(target.registered.has("list")).toBe(false);
  });

  test("unregisters the previous node when a dynamic path changes", () => {
    const target = createTarget();
    const store = new TestStore({ boardId: "board-1", cardCount: 3 });

    exposeStore(
      target,
      (state) => `boards/${state.boardId}`,
      store,
      (state) => ({
        type: "view",
        props: { card_count: state.cardCount },
      }),
    );

    expect(target.registered.has("boards/board-1")).toBe(true);

    store.setState({ boardId: "board-2", cardCount: 5 });

    expect(target.registered.has("boards/board-1")).toBe(false);
    expect(target.registered.get("boards/board-2")).toEqual({
      type: "view",
      props: { card_count: 5 },
    });
    expect(target.unregistered).toContainEqual({ path: "boards/board-1", opts: { recursive: true } });
  });

  test("skips projection work when equality reports no relevant change", () => {
    const target = createTarget();
    const store = new TestStore({ visible: 1, ignored: "a" });
    let projectCount = 0;

    exposeStore(
      target,
      "view",
      store,
      (state) => {
        projectCount++;
        return { type: "view", props: { visible: state.visible } };
      },
      { equals: (previous, next) => previous.visible === next.visible },
    );

    store.setState({ visible: 1, ignored: "b" });
    expect(projectCount).toBe(1);
    expect(target.registered.get("view")).toEqual({
      type: "view",
      props: { visible: 1 },
    });

    store.setState({ visible: 2, ignored: "b" });
    expect(projectCount).toBe(2);
    expect(target.registered.get("view")).toEqual({
      type: "view",
      props: { visible: 2 },
    });
  });

  test("debounces rapid store emissions before recomputing the descriptor", async () => {
    const target = createTarget();
    const store = new TestStore({ count: 0 });
    const projected: number[] = [];

    const cleanup = exposeStore(
      target,
      "counter",
      store,
      (state) => {
        projected.push(state.count);
        return { type: "status", props: { count: state.count } };
      },
      { debounceMs: 10 },
    );

    store.setState({ count: 1 });
    store.setState({ count: 2 });

    expect(projected).toEqual([0]);

    await sleep(20);

    expect(projected).toEqual([0, 2]);
    expect(target.registered.get("counter")).toEqual({
      type: "status",
      props: { count: 2 },
    });

    cleanup();
  });

  test("supports subscriptions with an unsubscribe method", () => {
    const target = createTarget();
    let unsubscribed = false;

    const store: StateStore<{ ready: boolean }> = {
      getState: () => ({ ready: true }),
      subscribe: () => ({
        unsubscribe() {
          unsubscribed = true;
        },
      }),
    };

    const cleanup = exposeStore(target, "status", store, (state) => ({
      type: "status",
      props: { ready: state.ready },
    }));

    cleanup();

    expect(unsubscribed).toBe(true);
  });
});

File: packages/go/slop-ai/store_test.go
package slop

import (
	"sync"
	"testing"
)

type testStoreState struct {
	Board   string
	Count   int
	Ignored string
}

type testStore struct {
	mu        sync.Mutex
	state     testStoreState
	listeners map[int]func()
	nextID    int
}

func newTestStore(state testStoreState) *testStore {
	return &testStore{state: state, listeners: map[int]func(){}}
}

func (s *testStore) GetState() testStoreState {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.state
}

func (s *testStore) Subscribe(listener func()) StoreUnsubscribe {
	s.mu.Lock()
	defer s.mu.Unlock()

	id := s.nextID
	s.nextID++
	s.listeners[id] = listener

	return func() {
		s.mu.Lock()
		defer s.mu.Unlock()
		delete(s.listeners, id)
	}
}

func (s *testStore) SetState(state testStoreState) {
	s.mu.Lock()
	s.state = state
	listeners := make([]func(), 0, len(s.listeners))
	for _, listener := range s.listeners {
		listeners = append(listeners, listener)
	}
	s.mu.Unlock()

	for _, listener := range listeners {
		listener()
	}
}

func (s *testStore) ListenerCount() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return len(s.listeners)
}

func TestExposeStoreRegistersAndUpdates(t *testing.T) {
	server := NewServer("app", "App")
	store := newTestStore(testStoreState{Count: 1})

	ExposeStore(server, "counter", store, func(state testStoreState) Node {
		return Node{Type: "status", Props: Props{"count": state.Count}}
	})

	tree := server.Tree()
	if got := tree.Children[0].Properties["count"]; got != 1 {
		t.Fatalf("expected count 1, got %v", got)
	}

	store.SetState(testStoreState{Count: 2})

	tree = server.Tree()
	if got := tree.Children[0].Properties["count"]; got != 2 {
		t.Fatalf("expected count 2, got %v", got)
	}
}

func TestExposeStoreCleanupUnsubscribesAndUnregistersRecursively(t *testing.T) {
	server := NewServer("app", "App")
	store := newTestStore(testStoreState{Count: 1})
	server.Register("counter/details", Node{Type: "group"})

	dispose := ExposeStore(server, "counter", store, func(state testStoreState) Node {
		return Node{Type: "status", Props: Props{"count": state.Count}}
	})

	if got := store.ListenerCount(); got != 1 {
		t.Fatalf("expected one listener, got %d", got)
	}

	dispose()

	if got := store.ListenerCount(); got != 0 {
		t.Fatalf("expected no listeners, got %d", got)
	}
	if len(server.Tree().Children) != 0 {
		t.Fatalf("expected empty tree after cleanup")
	}

	store.SetState(testStoreState{Count: 2})
	if len(server.Tree().Children) != 0 {
		t.Fatalf("expected cleanup to prevent later updates")
	}
}

func TestExposeStoreDynamicPathAndEquals(t *testing.T) {
	server := NewServer("app", "App")
	store := newTestStore(testStoreState{Board: "one", Count: 1, Ignored: "a"})
	projectCount := 0

	ExposeStore(
		server,
		"boards/one",
		store,
		func(state testStoreState) Node {
			projectCount++
			return Node{Type: "view", Props: Props{"count": state.Count}}
		},
		WithStorePath(func(state testStoreState) string {
			return "boards/" + state.Board
		}),
		WithStoreEquals(func(previous, next testStoreState) bool {
			return previous.Board == next.Board && previous.Count == next.Count
		}),
	)

	store.SetState(testStoreState{Board: "one", Count: 1, Ignored: "b"})
	if projectCount != 1 {
		t.Fatalf("expected equality skip to avoid projection, got %d projections", projectCount)
	}

	store.SetState(testStoreState{Board: "two", Count: 3, Ignored: "b"})

	tree := server.Tree()
	boards := tree.Children[0]
	if boards.ID != "boards" || boards.Children[0].ID != "two" {
		t.Fatalf("expected boards/two path, got %#v", boards)
	}
	if got := boards.Children[0].Properties["count"]; got != 3 {
		t.Fatalf("expected count 3, got %v", got)
	}
}
